Apply Fixed Code of Smartfs journaling

kavyasrinunna opened the pull request:
Signed-off-by: Kavya <kavya.nunna@partner.samsung.com>

sunghan-chang commented on os/fs/driver/mtd/smart.c:
```diff
@@ -195,7 +210,6 @@
  * 7:   0000       15:  0101
  */
 
-#ifdef CONFIG_MTD_SMART_WEAR_LEVEL
```
Why do you remove this conditional?

sunghan-chang commented on os/fs/driver/mtd/smart.c:
```diff
@@ -204,16 +218,13 @@ static const uint8_t gWearLevelToBitMap4[] = {
 
 /* Map a Wear Level bit pattern back to the wear level */
 
-#ifdef CONFIG_MTD_SMART_WEAR_LEVEL
```
same as previous

Taejun-Kwon commented:
I applied code with several commits.
Do not push whole code at the same time.
Please seperate them as I pushed in internal git.

Taejun-Kwon commented on os/fs/smartfs/smartfs_smart.c:
```diff
@@ -1526,23 +1526,31 @@ static int smartfs_bind(FAR struct inode *blkdriver, const void *data, void **ha
+		fdbg("recovery failed!!\n");
+	}
+#endif
+
```
I gave you this as a temporarily solution, below is more good for readability.

#ifdef CONFIG_SMARTFS_SECTOR_RECOVERY
         ret = smartfs_recover(fs);
         if (ret != 0) {
               fdbg();
               goto error_with_semaphore;
         }
#endif
         smartfs_semgive(fs);
         return OK;

error_with_semaphore:
         smartfs_semgive(fs);
          kmm_free(fs);

kavyasrinunna commented:
I have addressed review comments Kindly review all patches: https://github.com/Samsung/TizenRT/pull/524/commits/4eea7508ce2ee9e9429c039f5fd1c7d87a9f6584

olegartys commented on os/fs/driver/mtd/smart.c:
```diff
@@ -3558,23 +3566,68 @@ static uint16_t smart_findfreephyssector(FAR struct smart_struct_s *dev, uint8_t
-				physicalsector = x;
-				dev->lastallocblock = allocblock;
-				break;
+				ret = MTD_READ(dev->mtd, readaddr, dev->mtdBlksPerSector * dev->geo.blocksize, sector_buff);
```
Do not we need to check 'ret' after MTD_READ?

kavyasrinunna replied:
Yes, we need to check ret . It is added in https://github.com/Samsung/TizenRT/pull/524/commits/1a45d0e12b357d26bb843cada4b4a856325d9950

Taejun-Kwon commented on os/fs/driver/mtd/smart.c:
```diff
@@ -3317,14 +3386,15 @@ static uint16_t smart_findfreephyssector(FAR struct smart_struct_s *dev, uint8_t
-#ifdef CONFIG_MTD_SMART_WEAR_LEVEL
 retry:
-#endif
+	bitflipped = FALSE;
```
This should be added in internal git too.

kavyasrinunna replied:
Okay

Taejun-Kwon commented on os/fs/driver/mtd/smart.c:
```diff
@@ -3489,23 +3564,73 @@ static uint16_t smart_findfreephyssector(FAR struct smart_struct_s *dev, uint8_t
+		if (bitflipped) {
+			fdbg("retry allocation \n");
+			canrelocate = FALSE;
+			goto retry;
```
Now I can see bug in my code.
If we move to retry here or below, there's leakage because we didn't free sector_buff.
Checking condition or freeing code should be added...

kavyasrinunna replied:
Yes, It is added in https://github.com/Samsung/TizenRT/pull/524/commits/592e0aaea6a6fbbc99347cd3cc7ffce7b044bffd

pradeepns85 commented on os/fs/driver/mtd/smart.c:
```diff
@@ -3489,23 +3564,73 @@ static uint16_t smart_findfreephyssector(FAR struct smart_struct_s *dev, uint8_t
 	}
 
+error:
 	if (physicalsector == 0xFFFF) {
```
I think You can club error cases and remove the redundant code.

if ((physicalsector == 0xFFFF) || (physicalsector > XYZ) {
            kmm_free();
}

kavyasrinunna replied:
yes, it is done in https://github.com/Samsung/TizenRT/pull/524/commits/592e0aaea6a6fbbc99347cd3cc7ffce7b044bffd
